### PR TITLE
836 public search option

### PR DESCRIPTION
**Pull request by tcnichol**
This pull request will enable a search of public data for non-logged in users.

It also fixes a problem. When a logged in user would search, public and authenticated matches were not being shown. In the query 'public' and 'authenticated' had to be in all lowercase. Additionally, not all files were having their status updated when the dataset changed. This is also fixed.

To test, create datasets. Make sure some are public and authenticated.

If no user is logged in, public datasets and files should show up in search results. 

If a user is logged in, public and authenticated datasets and files should show up in search results, along with their own items or datasets explicitly shared. 

Changing the status of a dataset should be reflected in search results.

### CODE REVIEW

**Review by ddey2 — CHANGES_REQUESTED** [review]
@tcnichol I can see the public datasets in the dropdown list for public search. But when I click enter or click on any of the dataset form the dropdown list, it doesn't take me to the search result page or the dataset page itself.
Screenshot 2024-03-11 at 3 31 35 PM

**Comment by tcnichol** [comment]
> @tcnichol I can see the public datasets in the dropdown list for public search. But when I click enter or click on any of the dataset form the dropdown list, it doesn't take me to the search result page or the dataset page itself. Screenshot 2024-03-11 at 3 31 35 PM

I think I have fixed this. I noticed that in the search results the links were not for public_datasets or public_files, they were the same as for logged in search. I added a PublicSearchResult component which should fix these issues.

File: frontend/src/components/search/PublicSearch.tsx [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+									onChange={() => {
+										setLuceneOn((prevState) => !prevState);
+									}}
+									name="Query String"
+								/>
+							}
+							label={
+								<Typography
+									variant="body1"
+									sx={{
```
Comment (ddey2) [inline frontend/src/components/search/PublicSearch.tsx:47]: Can you put the styles in main.css, may be?

File: frontend/src/components/search/PublicSearchResult.tsx [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+}
+
+export function PublicSearchResult(props) {
+	const { data } = props;
+
+	if (data.length > 0) {
+		return (
+			<List
+				sx={{
+					width: "100%",
```
Comment (ddey2) [inline frontend/src/components/search/PublicSearchResult.tsx:87]: here too

File: frontend/src/components/search/PublicSearchResult.tsx [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+	return (
+		<>
+			<ListItemAvatar sx={{ color: theme.palette.primary.main }}>
+				<InsertDriveFileIcon />
+			</ListItemAvatar>
+			<Box sx={{ marginTop: "5px" }}>
+				<MuiLink
+					component={Link}
+					to={`/public_files/${item._id}?public_dataset=${item.dataset_id}`}
+					sx={{ fontWeight: "bold", fontSize: "18px" }}
```
Comment (ddey2) [inline frontend/src/components/search/PublicSearchResult.tsx:65]: here too.

File: frontend/src/components/search/PublicSearchResult.tsx [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+	return (
+		<>
+			<ListItemAvatar sx={{ color: theme.palette.primary.main }}>
+				<DatasetIcon />
+			</ListItemAvatar>
+			<Box sx={{ marginTop: "5px" }}>
+				<MuiLink
+					component={Link}
+					to={`/public_datasets/${item._id}`}
+					sx={{ fontWeight: "bold", fontSize: "18px" }}
```
Comment (ddey2) [inline frontend/src/components/search/PublicSearchResult.tsx:40]: here too...looks like you're using the same styling properties again.

**Review by ddey2 — CHANGES_REQUESTED** [review]

File: frontend/src/components/search/PublicSearchResult.tsx (hunk 4 above)
Comment (tcnichol) [reply frontend/src/components/search/PublicSearchResult.tsx:40]: I created a separate issue for these.

**Comment by tcnichol** [comment]
@longshuicy 

I checked and the search result links should be fixed for public search now, for both files and datasets.

File: frontend/src/components/PublicLayout.tsx [hunk 5]
```diff
@@ -235,6 +235,18 @@ export default function PersistentDrawerLeft(props) {
 						</ListItemButton>
 					</ListItem>
 				</List>
+				<Divider/>
+				<List>
+					<ListItem key={"public_search"} disablePadding>
+						<ListItemButton component={RouterLink} to="/public_search">
+							<ListItemIcon>
+								<SearchDatasetIcon />
```
Comment (ddey2) [inline frontend/src/components/PublicLayout.tsx:243]: It gives error here as you never imported SearchDatasetIcon in this file.

**Review by ddey2 — CHANGES_REQUESTED** [review]
Rest looks good to me. Once you import SearchDatasetIcon, public search works.

File: frontend/src/components/PublicLayout.tsx (hunk 5 above)
Comment (longshuicy) [reply frontend/src/components/PublicLayout.tsx:243]: @tcnichol is this addressed?

Comment (tcnichol) [reply frontend/src/components/PublicLayout.tsx:244]: Just pushed a fix for this.

**Comment by ddey2** [comment]
@tcnichol Clicking on the link after the search doesn't work. Clicking on the dataset takes you to the main page.
Screenshot 2024-04-16 at 2 39 53 PM

**Review by ddey2 — CHANGES_REQUESTED** [review]
left comment

**Comment by tcnichol** [comment]
@ddey2 I noticed that there was an inconsistency in the routes for public_datasets. Some spots had public/datasets instead. Went through and I think I fixed that.

Though now the link to the public search is broken at the top of the Public page, working on that.

**Review by ddey2 — APPROVED** [review]
Looks good.

File: backend/app/routers/elasticsearch.py [hunk 6]
```diff
@@ -23,8 +23,8 @@ def _add_permissions_clause(
             "should": [
                 {"term": {"creator": username}},
                 {"term": {"user_ids": username}},
-                {"term": {"status": "AUTHENTICATED"}},
-                {"term": {"status": "PUBLIC"}},
+                {"term": {"status": "authenticated"}},
+                {"term": {"status": "public"}},
```
Comment (longshuicy) [inline backend/app/routers/elasticsearch.py:27]: Just curious, why capitalized doesn't work? In the database and in the router logic, it's all capitalized?

File: frontend/src/routes.tsx [hunk 7]
```diff
@@ -194,7 +195,7 @@ export const AppRoutes = (): JSX.Element => {
 						</PrivateRoute>
 					}
 				/>
-				<Route path="/public/files/:fileId" element={<PublicFileComponent />} />
+				<Route path="/public_files/:fileId" element={<PublicFileComponent />} />
```
Comment (longshuicy) [inline frontend/src/routes.tsx:198]: Are we using public/files or public_files? 
In other PRs that got merged in, I fixed places that are using "_"...

Comment (longshuicy) [reply frontend/src/routes.tsx:198]: If you would like to use "_", you might have to 1. merge the latest main, 2. do a project wide search of places still using /
E.g. If you go to a file under public dataset, it right now still goes to /public/file and hence broken.

**Review by longshuicy — CHANGES_REQUESTED** [review]
Public search seems to be able to match dataset, not files though. Might be your clause is specifically matching dataset status. Not sure how to get around, but at least I would change the label `Search for Datasets and Files`
image

File: frontend/src/routes.tsx (hunk 7 above)
Comment (tcnichol) [reply frontend/src/routes.tsx:198]: I think I got all of these fixed with a merge and a search/replace. Testing.

Comment (tcnichol) [reply frontend/src/routes.tsx:198]: I also think that files should have a status. Checking the populate fake data, the status for files is 'null' in some cases and those should match the datasets.

File: backend/app/routers/elasticsearch.py (hunk 6 above)
Comment (tcnichol) [reply backend/app/routers/elasticsearch.py:27]: I have fixed the problem with the files having null status. They now should always have the same status as the dataset.

I'm still not getting files in the search results for public search, looking into this.

File: frontend/src/routes.tsx (hunk 7 above)
Comment (longshuicy) [reply frontend/src/routes.tsx:198]: Thank you. This has been resolved.

File: backend/app/routers/elasticsearch.py (hunk 6 above)
Comment (longshuicy) [reply backend/app/routers/elasticsearch.py:27]: I don't think this has been resolved yet?

**Review by longshuicy — CHANGES_REQUESTED** [review]

File: backend/app/routers/elasticsearch.py (hunk 6 above)
Comment (tcnichol) [reply backend/app/routers/elasticsearch.py:27]: I think this is fixed. Files should be showing up in public search now.

**Comment by longshuicy** [comment]
@tcnichol any chance you have addressed public files not searchable issue? If it is challenging, could you create another issue to address that separately? The rest of this PR looks good.

**Comment by tcnichol** [comment]
> @tcnichol any chance you have addressed public files not searchable issue? If it is challenging, could you create another issue to address that separately? The rest of this PR looks good.

Let me check. I think it should be fixed because the problem was that the fields for public or authenticated were null for files in elasticsearch, but that should be fixed now. I'll make sure it is working.

**Comment by lmarini** [comment]
> > @tcnichol any chance you have addressed public files not searchable issue? If it is challenging, could you create another issue to address that separately? The rest of this PR looks good.
> 
> Let me check. I think it should be fixed because the problem was that the fields for public or authenticated were null for files in elasticsearch, but that should be fixed now. I'll make sure it is working.

@tcnichol is this fixed now? Thanks.

**Review by longshuicy — APPROVED** [review]
After drop volume and restart, elasticsearch is indexing properly and I can search for files. Approved.

**Pull request merged**